### DISCUSSION TITLE
Add regexp into pdfsnip rule to remove extraspaces and linebreaks from snips from pdfs

### DISCUSSION THREAD

**Discussion by arunnbabu81** [buggyj/tiddlyclip-plugin#121] [Q&A] [answered]
I use this [editor-toolbar button](https://talk.tiddlywiki.org/t/editor-toolbar-button-to-remove-extra-space-and-line-breaks/6873) shared by etardiff in the talk tiddlywiki forum to remove extraspaces and linebreaks between the snips taken from some pdfs. 

Before removing the extraspaces and linebreaks it looks like this.

<img width="675" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/55e1e9f6-2f5f-49ad-8773-600b1b2cfbe2">

After removing the extraspaces and linebreaks using the editor toolbar button, it looks like this.

<img width="789" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/1e45bf0b-ee24-4c2f-834e-09cb9ebfc882">

Is there any way to add this regex pattern to remove extraspaces and linebreaks  into the [pdfsnip rule](https://sidetabs-wiki.tiddlyhost.com/#LocalmpdfFieldModifiers)

<img width="782" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/b771c743-d8f7-47cd-852f-ec3a736fcf3e">

**Comment by buggyj** [comment]
Try this macro, like: `((*@reduceLines(@text)*))`


[reduceLines.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/11517546/reduceLines.js.json.zip)

**Reply by arunnbabu81** [reply]
> Try this macro, like: `((*@reduceLines(@text)*))`
> 
> [reduceLines.js.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/11517546/reduceLines.js.json.zip)

Thanks @buggyj for this. I am traveling now. Will check and respond tomorrow.

**Comment by arunnbabu81** [comment]
> Before removing the extraspaces and linebreaks it looks like this.
> <img alt="image" width="675" src="https://user-images.githubusercontent.com/67494083/239369303-55e1e9f6-2f5f-49ad-8773-600b1b2cfbe2.png">

I am not able to reproduce behaviour shown in the picture above. 
What I get on snip from pdf now is like given below

<img width="827" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/63ee264b-10f7-45e2-b3d5-d3503d673eb9">


But after adding the reduceLines json to my wiki and changing the pdf rule to this 

<img width="841" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/1aa2280a-0979-45f9-9282-0711b07ddc6b">

What I get on pdf snip is 

<img width="827" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/af4a1641-5b67-411b-a47e-4b7faed577ae">

What I actaully want is add a line break after each punctuation mark or full-stop like shown in the image below. 

<img width="830" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/ea8f233b-3c5b-47c0-976b-9f9b7a0c0c22">

If I go further, is it possible to add a bullet formatting to each sentence like shown below on snip.

<img width="841" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/e2d233ca-3b03-4884-a8c5-cfbf00c9a030">

Is it possible to do this as two separate rules so that I can use them as I need.

**Comment by buggyj** [comment, marked as answer]
you can try this:

```
#extraInsert=*
@text=((*#extraInsert*))((*@removelf(@text)*))
#rule=/(\.)/g
#term=$1\n((*#extraInsert*))
@text=((*@text/#rule/#term*))
$text=((*@text*))\n\n
```

**Reply by arunnbabu81** [reply]
> you can try this:
> 
> ```
> #extraInsert=*
> @text=((*#extraInsert*))((*@removelf(@text)*))
> #rule=/(\.)/g
> #term=$1\n((*#extraInsert*))
> @text=((*@text/#rule/#term*))
> $text=((*@text*))\n\n
> ```

This one works.
How to modify this code to get each sentence limited by full stop as separate lines without adding the bullet formatting as shown below.
 <img alt="image" width="830" src="https://user-images.githubusercontent.com/67494083/239698396-ea8f233b-3c5b-47c0-976b-9f9b7a0c0c22.png">

**Comment by buggyj** [comment]
change

#extraInsert=

**Comment by arunnbabu81** [comment]
```
#extraInsert=((*@userstring*))
@text=((*#extraInsert*))((*@removelf(@text)*))
#rule=/(\.)/g
#term=$1\n((*#extraInsert*))
@text=((*@text/#rule/#term*))
$text=((*$text*))\n\n((*@text*))\n\n
```

This is the rule I am using now. I added `#extraInsert=((*@userstring*)) `which is more flexible and can be changed at the time of the clipping. Also I modified the last line as `$text=((*$text*))\n\n((*@text*))\n\n `so that the subsequent clips are not getting deleted (based on your old suggestion).

I have one doubt. If I keep the userstring as empty, I get this result

<img width="1795" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/6cde1940-e5e6-42d9-b121-5c31a1f6202e">


How to get an extra line break after each sentence as shown in the image below `using userstring input with the abovementioned rule`. 

<img width="1797" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/950e2131-8628-4d41-916e-cb7084de4b0f">

Is that possible ? It's not an urgent need. Just a doubt.

**Reply by buggyj** [reply]
try entering `\n`

**Reply by arunnbabu81** [reply]
<img width="1697" alt="image" src="https://github.com/buggyj/tiddlyclip-plugin/assets/67494083/448afe32-2304-4c54-9c0b-4b60c273bf42">

Did you meant using `\n `as userstring as shown in the image above. If yes, it's not working

**Reply by buggyj** [reply]
it needs converting to a 'linefeed'

try

```
#nl=\n
#n=n
#nlescaped=\((*#n*))
@userstring=((*#nlescaped==@userstring*??*#nl*??*@userstring*))
```

**Reply by arunnbabu81** [reply]
That was awesome. It works perfectly. Note taking has become so easy with all these new additions to tiddlyclip. Whenever I ask some doubts you come up with new tricks. 

I have seen that you have updated the tiddlyclip demo page with new contents. Should I share some of the rules which I use (obviously based on your advice and suggestions) to be added to the demo page ?

**Reply by buggyj** [reply]
It is difficult to write good documentation. I think the first contact needs to be very simple and then slowly get more complicated. For the more complex stuff (the interesting stuff) maybe a separate TW is needed. 

There will be a new version of the plugin soon (version 3.0.x).
